include/rmm/device_buffer.hpp

```diff
@@ -259,6 +267,7 @@ class device_buffer {
   {
     set_stream(stream);
     if (new_capacity > capacity()) {
+      cuda_set_device_raii dev{_device};
       auto tmp            = device_buffer{new_capacity, stream, _mr};
```

Comment:
**question:** Does the appearance of this pattern suggest that the `device_buffer` constructor should have an (optional) `device` argument that one can provide, rather than relying on the implicit current cuda device (which is then managed by this raii object here)?

Reply:
If we did that, then if we are eliminating the `cuda_set_device_raii` here, then the constructor would have to first call `cudaSetDevice(device)`, and I assume it would do so using `cuda_set_device_raii`, which means on exiting the ctor the previous device would be reset (if different).

So then we would need to call `cuda_set_device_raii` again *after* calling the constructor with the optional `device` argument because of the subsequent `cudaMemcpyAsync`.  That could mean two calls to `cudaGetDevice` and four calls to `cudaSetDevice`, worst case. The way it is now, there is at most 1 `cudaGetDevice` and at most 2 `cudaSetDevice`.

Reply:
Hmm, my understanding from the docs was that runtime calls (excepting [some] of those to do with events, where the call has to happen with the live device matching the event's stream) don't care about the current device and hence allocation/deallocation (which, with a pool mr record events) are the only places we need to handle it.